### DISCUSSION TITLE
How to Read DXF file with acadsharp?

### DISCUSSION THREAD

**Discussion by farshadvalizade** [DomCR/ACadSharp#354] [Q&A]
hi guys . I have used this code but get error : object references not set ....

using (DxfReader reader = new DxfReader(file, null))
{
    CadDocument doc = reader.Read();
}

why? and what should I do?

**Comment by DomCR** [comment]
Hi @farshadvl,

This seems an error in the Reader, could you provide more information about the exception?
If you could provide the file that caused the error that would help a lot.

**Reply by farshadvalizade** [reply]
hi
this is error result :
+		$exception	{"Object reference not set to an instance of an object."}	System.NullReferenceException
+		Data	{System.Collections.ListDictionaryInternal}	System.Collections.IDictionary {System.Collections.ListDictionaryInternal}

		StackTrace	"   at ACadSharp.CadDocument.RegisterCollection[T](IObservableCollection`1 collection)\r\n   at ACadSharp.IO.CadDocumentBuilder.RegisterTables()\r\n   at ACadSharp.IO.DXF.DxfDocumentBuilder.BuildDocument()\r\n   at ACadSharp.IO.DxfReader.Read()\r\n   at KianSazehPiping.Forms.App.Form_DXF.button1_Click_1(Object sender, EventArgs e) in D:\\Source\\KianSazehPiping\\KianSazehPiping\\Forms\\App\\DXF.cs:line 1186"	string

+		TargetSite	{Void RegisterCollection[T](ACadSharp.IObservableCollection`1[T])}	System.Reflection.MethodBase {System.Reflection.RuntimeMethodInfo}

**Reply by DomCR** [reply]
Looking at the exception seems that one of the main tables in the document was null, this is usually caused because the version of the DXF is equal or lower than AC1009, there is a branch in development to fix this issue.

In the mean time you can try to use the methods `ReadEntities` or `ReadTables` to try to get the information that you want.

Hope this helps.

**Reply by farshadvalizade** [reply]
tnx man. I use nugetpackage and If you update that package I really appreciate it.

**Reply by DomCR** [reply]
I did a release 4 days ago, check that you have the latest version.

**Comment by farshadvalizade** [comment]
Hi DomCR I have update package to latest version  ACadSharp -Version 2.3.1-beta
again it says object refrence.... when read dxf file :
CadDocument doc = DxfReader.Read(file);

can u check it.my dxf file here I have uploaded
[4-FUG-16802-AAR61-N_Sht_1.zip](https://github.com/user-attachments/files/15750576/4-FUG-16802-AAR61-N_Sht_1.zip)

**Reply by DomCR** [reply]
Hi @farshadvl 

I've tested the file, it seems that is not a regular file, it does not have a version assigned and all the subclasses are missing, I would suggest to use the `ReadEntities` or `ReadTables` instead of the regular `Read()` method.

This PR may be able to fix your issue, the AC1009 version is quite similar to your file structure.
- https://github.com/DomCR/ACadSharp/pull/165

**Reply by farshadvalizade** [reply]
can you explain an example how to use ReadEntitiy?
How can I load dxf file into CadDocument an dafter that ReadEntities?

**Reply by DomCR** [reply]
I've added examples on how can you read the entities 
https://github.com/DomCR/ACadSharp/blob/bdafc5f5410ab503b8aea3652c5376fb8f6a8754/ACadSharp.Examples/ReaderExamples.cs#L40

I've tested the method with your file, unfortunately the file does not follow the proper DXF format specified in the documentation and the polylines are not read, this is not the first time that I see this kind of dxf so I'll try to find a fix for this issue.

**Reply by DomCR** [reply]
- https://github.com/DomCR/ACadSharp/issues/359

**Reply by DomCR** [reply]
Out of curiosity, could you provide the source of the file? I mean if is a public DXF generator that I can use to run some test.

**Reply by farshadvalizade** [reply]
Tnx man . I appreciate your response.
this file was generated from PDMS software.  this DXF is output from this application.

**Comment by farshadvalizade** [comment]
Hi Dom I have read this dxf with this library:IxMilia
this is my code:
`                        DxfFile dxfFile = DxfFile.Load(file);

                        if (dxfFile != null)
                        {
                            foreach (DxfEntity entity in dxfFile.Entities)
                            {
                                if (entity.EntityType == DxfEntityType.Text)
                                {
                                    var text = (DxfText)entity;
                                    richTextBox1.Text += "[" + text.Location + "]" + text.Value + Environment.NewLine;
                                }
                            }

                        }`

**Reply by DomCR** [reply]
Hi @farshadvl,

I haven't used the library from IxMilia, but the latest master branch of `ACadSharp` should be able to read the file that you provided, here is an equivalent code for what you want to achieve.

```C#
	CadDocument doc = DxfReader.Read(test, this.onNotification);
	foreach (TextEntity text in doc.Entities.OfType<TextEntity>())
	{
		//Do something for each text entity
		text.Value = "New text value";
	}
```

If you are using the Nuget package let me know and I'll release an update.

Hope this helps.
